File: cloud-upload-check.py
"""
Cloud upload file check
-----------------------
A file inspection utility that compares file names and sizes between the input (source) directory and the Google Cloud Project (GCP) storage bucket using the same configuration YAML file as the original script.

Usage:
    python cloud-upload-check.py path/to/name-of-configuration-file.yml

Author:  matt.grossi@noaa.gov with creation and refactoring assistance from
         Google Gemini Coding Partner
Project: Southeast Fishery Independent Survey (SEFIS)
Version: 2026.1.0
Note:    Gemini Coding Partner was used to assist with developing this code.
         The code has been reviewed, edited, validated, and documented by NOAA
         Fisheries staff.
"""

# =============================================================================
# PACKAGE DEPENDENCIES
# =============================================================================

from urllib.parse import urlparse
import subprocess
import yaml
import sys
import os
import csv
import shutil
import argparse

# =============================================================================
# HELPER FUNCTIONS AND METHODS
# =============================================================================

def parse_args():
    """Parses command-line arguments."""
    parser = argparse.ArgumentParser(description="GoPro Clip-and-Stitch Utility")
    parser.add_argument("config_path", type=str, nargs="?",
        default="configurations.yml",
        help="Path to the YAML configuration file (default: configurations.yml)"
    )
    return parser.parse_args()

def load_config(config_path: str = 'configurations.yml'):
    """Loads and verifies the YAML configuration file.
    
    Arguments
    ---------
    config_path (str): The file path to the YAML configuration file. Defaults
            to 'configurations.yml'.
    Returns
    -------
    dict
    """
    with open(config_path, 'r') as f:
        config = yaml.safe_load(f)

    # Validate keys
    REQUIRED_KEYS = {'output_directory', 'gcp_bucket_path'}
    missing = [f"  - '{k}'" for k in REQUIRED_KEYS if k not in config]
    if missing:
        error_msg = (
            "\n[!] CONFIGURATION ERROR: Missing required keys in YAML:\n" +
            "\n".join(missing) + "\n\n"
        )
        raise ValueError(error_msg)
    config['video_extension'] = config.get('video_extension', '.MP4')

    return config

def extract_gcp_prefix(bucket_path):
    """
    Safely extracts the folder prefix from a GCP bucket path, automatically
    stripping out wildcards, file names, or extensions.

    Arguments
    ---------
    bucket_path (str): full path to GCP storage bucket and folder

    Returns
    -------
    Returns the file prefix only
    """
    # Parse the gs:// URI safely (separates bucket from path)
    # urlparse("gs://my-bucket/folder/*.MP4").path -> "/folder/*.MP4"
    parsed_path = urlparse(bucket_path).path
    
    # Strip the leading slash left over by urlparse
    prefix = parsed_path.lstrip('/')
    
    # Return with a trailing slash, or empty string if it's the bucket root
    return f"{prefix.rstrip('/')}/" if prefix else ""

def get_cloud_manifest(bucket_path, extension=None):
    """Queries GCP bucket and returns data in a dictionary.
    
    Arguments
    ---------
    bucket_path (str): full path to GCP storage bucket and folder to check
    extension (str): file extension to filter by (optional)

    Returns
    -------
    dict of file names and sizes 
    
    """
    bucket_path = f"{bucket_path.rstrip('/*')}/*"
    print(f"Fetching cloud bucket inventory from {bucket_path}...")
    # Use gcloud CLI to retrieve names and sizes of all objects in the bucket
    # Use shutil.which to find the actual path of gcloud (handles .cmd on Windows)
    gcloud_exec = shutil.which("gcloud")
    
    if not gcloud_exec:
        print("\nERROR: 'gcloud' command not found. Is Google Cloud SDK installed and in your PATH?")
        return False
    cmd = [
        gcloud_exec, "storage", "objects", "list", 
        bucket_path, 
        '--format=csv[no-heading](name, size)'
    ]
    
    # Run gcloud and capture stdout directly into memory
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
    except subprocess.CalledProcessError as e:
        print(f"\n❌ ERROR running gcloud command: {e.stderr}")
        sys.exit(1)
    except FileNotFoundError:
        print("\n❌ ERROR: 'gcloud' CLI tool not found. Make sure it's installed and in your PATH.")
        sys.exit(1)

    # Parse the text string using Python's csv reader and store as dict
    cloud_data = {}
    ext_lower = f".{extension.lstrip('.').lower()}" if extension else None
    reader = csv.reader(result.stdout.strip().splitlines())
    for row in reader:
        if row:
            path, size = row[0].strip(), row[1].strip()
        
            if not ext_lower or path.lower().endswith(ext_lower):
                cloud_data[path] = int(size)
            
    return cloud_data

def get_local_manifest(local_path, prefix, extension=None):
    """Scans the local top-level directory and returns data in a dictionary.
    
    Arguments
    ---------
    local_path (str): local directory to compare to cloud
    prefix (str): prefix appended to file name in the storage bucket
    extension (str): file extension to filter by (optional)
    """
    print(f"Scanning local files in {local_path}...")
    
    if not os.path.exists(local_path):
        print(f"\n❌ ERROR: Cannot access local path: {local_path}")
        sys.exit(1)

    ext_lower = extension.lower() if extension else None    
    local_data = {}
    for file in os.listdir(local_path):
        full_path = os.path.join(local_path, file)
        
        if os.path.isfile(full_path) and (not ext_lower or file.lower().endswith(ext_lower)):
            try:
                size = os.path.getsize(full_path)
                # Format to match the cloud path structure
                gcp_style_path = f"{prefix}{file}".replace("//", "/")
                local_data[gcp_style_path] = size
            except Exception as e:
                print(f"  ⚠️ Error reading size for {file}: {e}")
                
    return local_data

def compare_inventories(local, cloud):
    """Compares local file names and sizes with those in the cloud and prints
    results.
    
    Arguments
    ---------
    local (dict): dictionary containing `file_name: file_size` pairs for all
        local files
    cloud (dict): dictionary containing `file_name: file size` pairs for all
        cloud files    
    """

    print("\nAnalyzing discrepancies...")
    
    # Check for missing files in either location
    missing_in_cloud = [p for p in local if p not in cloud]
    missing_in_local = [p for p in cloud if p not in local]

    # Check for size mismatches
    size_mismatches = []
    for p, local_size in local.items():
        if p in cloud and cloud[p] != local_size:
            size_mismatches.append((p, local_size, cloud[p]))
    extension = p.split('.')[-1]

    print("\n=====================================")
    print("        COMPARISON RESULTS           ")
    print("=====================================")
    print(f"Local drive count: {len(local)} {extension} files")
    print(f"GCP Bucket count:  {len(cloud)} {extension} objects")
    print("-------------------------------------")

    if not missing_in_cloud and not missing_in_local and not size_mismatches:
        print("🎉 SUCCESS! Every file matches perfectly in name and size.")
    else:
        if missing_in_cloud:
            print(f"❌ Missing in Cloud ({len(missing_in_cloud)}):")
            for item in missing_in_cloud[:10]:
                print(f"  - {item}")
            if len(missing_in_cloud) > 10:
                print(f"  ... and {len(missing_in_cloud)-10} more.")
            
        if missing_in_local:
            print(f"⚠️ Extra in Cloud / Missing Locally ({len(missing_in_local)}):")
            for item in missing_in_local[:10]:
                print(f"  - {item}")
            if len(missing_in_local) > 10:
                print(f"  ... and {len(missing_in_local)-10} more.")
            
        if size_mismatches:
            print(f"❌ Byte Size Mismatches ({len(size_mismatches)}):")
            for item, l_sz, c_sz in size_mismatches[:10]:
                print(f"  - {item} (Local: {l_sz} bytes, Cloud: {c_sz} bytes)")
            if len(size_mismatches) > 10:
                print(f"  ... and {len(size_mismatches)-10} more.")

if __name__ == "__main__":
    # Load configuration file
    args = parse_args()
    config = load_config(args.config_path)

    # Fetch cloud and local inventory
    cloud_inventory = get_cloud_manifest(
        bucket_path=config['gcp_bucket_path'],
        extension=config['video_extension']
        )
    prefix = extract_gcp_prefix(config['gcp_bucket_path'])
    local_inventory = get_local_manifest(
        local_path=config['output_directory'],
        prefix=prefix,
        extension=config['video_extension']
        )
    
    # Compare and Print
    compare_inventories(local=local_inventory, cloud=cloud_inventory)

File: cloud-upload-check.R
#' Cloud upload file check (R Version)
#' -----------------------------------
#' A file inspection utility that compares file names and sizes between the
#' input (source) directory and the Google Cloud Project (GCP) storage bucket
#' using the same configuration YAML file as the original script.
#'
#' Author:  Southeast Fishery Independent Survey (SEFIS)
#' Version: 2026.1.1

# Ensure the required yaml package is available
if (!requireNamespace("yaml", quietly = TRUE)) {
  stop("The 'yaml' package is required. Please install it using: install.packages('yaml')")
}

# =============================================================================
# HELPER FUNCTIONS AND METHODS
# =============================================================================

parse_args <- function() {
  #' Parses command-line arguments or prompts the user via a file dialog window.
  args <- commandArgs(trailingOnly = TRUE)
  
  # If run via command line with an argument, use that argument
  if (length(args) > 0) {
    return(args[1])
  }
  
  # If run interactively (RStudio/Positron/Console), pop up a file explorer
  if (interactive()) {
    cat("\n[➔] A file dialog has opened. Please select your configuration YAML file...\n")
    flush.console() # Force text to display before the window blocks the console
    
    config_path <- tryCatch({
      file.choose()
    }, error = function(e) {
      cat("\n⚠️ File selection cancelled by user. Attempting default 'configurations.yml'...\n")
      return("configurations.yml")
    })
    
    return(config_path)
  }
  
  # Fallback default for headless/non-interactive environments
  return("configurations.yml")
}

load_config <- function(config_path = "configurations.yml") {
  #' Loads and verifies the YAML configuration file.
  if (!file.exists(config_path)) {
    stop(sprintf("\n[!] ERROR: Configuration file not found at '%s'\n", config_path))
  }
  
  config <- suppressWarnings(yaml::read_yaml(config_path))
  
  # Validate keys
  required_keys <- c("output_directory", "gcp_bucket_path")
  missing <- required_keys[!(required_keys %in% names(config))]
  
  if (length(missing) > 0) {
    missing_formatted <- paste(sprintf("  - '%s'", missing), collapse = "\n")
    stop(sprintf("\n[!] CONFIGURATION ERROR: Missing required keys in YAML:\n%s\n\n", missing_formatted))
  }
  
  if (is.null(config$video_extension)) {
    config$video_extension <- ".MP4"
  }
  
  return(config)
}

extract_gcp_prefix <- function(bucket_path) {
  #' Safely extracts the folder prefix from a GCP bucket path.
  prefix <- gsub("^gs://[^/]+/", "", bucket_path)
  prefix <- gsub("^/", "", prefix) 
  
  if (nchar(prefix) > 0) {
    clean_prefix <- gsub("[/*]+$", "", prefix)
    return(paste0(clean_prefix, "/"))
  } else {
    return("")
  }
}

get_cloud_manifest <- function(bucket_path, extension = NULL) {
  #' Queries GCP bucket using gcloud CLI and returns data in a named list.
  clean_path <- gsub("[/*]+$", "", bucket_path)
  gcloud_target_path <- paste0(clean_path, "/*")
  
  cat(sprintf("Fetching cloud bucket inventory from %s...\n", gcloud_target_path))
  
  gcloud_exec <- Sys.which("gcloud")
  if (gcloud_exec == "") {
    cat("\nERROR: 'gcloud' command not found. Is Google Cloud SDK installed and in your PATH?\n")
    return(FALSE)
  }
  
  cmd_args <- c(
    "storage", "objects", "list", 
    gcloud_target_path, 
    "--format=csv[no-heading](name,size)"
  )
  
  result <- system2(gcloud_exec, args = cmd_args, stdout = TRUE, stderr = TRUE)
  
  status <- attr(result, "status")
  if (!is.null(status) && status != 0) {
    cat(sprintf("\n❌ ERROR running gcloud command:\n%s\n", paste(result, collapse = "\n")))
    stop(status = 1)
  }
  
  cloud_data <- list()
  if (length(result) == 0 || (length(result) == 1 && result == "")) {
    return(cloud_data)
  }
  
  ext_lower <- if (!is.null(extension)) tolower(sub("^\\.*", ".", extension)) else NULL
  
  con <- textConnection(result)
  df <- tryCatch(
    read.csv(con, header = FALSE, stringsAsFactors = FALSE, colClasses = c("character", "numeric")),
    error = function(e) NULL
  )
  close(con)
  
  if (is.null(df) || nrow(df) == 0) return(cloud_data)
  
  for (i in 1:nrow(df)) {
    path <- trimws(df[i, 1])
    size <- df[i, 2]
    
    if (is.null(ext_lower) || endsWith(tolower(path), ext_lower)) {
      cloud_data[[path]] <- size
    }
  }
  
  return(cloud_data)
}

get_local_manifest <- function(local_path, prefix, extension = NULL) {
  #' Scans the local top-level directory and returns data in a named list.
  cat(sprintf("Scanning local files in %s...\n", local_path))
  
  if (!dir.exists(local_path)) {
    cat(sprintf("\n❌ ERROR: Cannot access local path: %s\n", local_path))
    stop(status = 1)
  }
  
  all_items <- list.files(local_path, full.names = FALSE, recursive = FALSE)
  ext_lower <- if (!is.null(extension)) tolower(sub("^\\.*", ".", extension)) else NULL
  
  local_data <- list()
  for (item in all_items) {
    full_path <- file.path(local_path, item)
    
    if (!dir.exists(full_path) && (is.null(ext_lower) || endsWith(tolower(item), ext_lower))) {
      tryCatch({
        size <- file.info(full_path)$size
        
        gcp_style_path <- paste0(prefix, item)
        gcp_style_path <- gsub("//", "/", gcp_style_path)
        
        local_data[[gcp_style_path]] <- size
      }, error = function(e) {
        cat(sprintf("  ⚠️ Error reading size for %s: %s\n", item, e$message))
      })
    }
  }
  
  return(local_data)
}

compare_inventories <- function(local, cloud, video_extension) {
  #' Compares names and sizes between both environments and handles empty edges.
  cat("\nAnalyzing discrepancies...\n")
  
  local_names <- names(local)
  cloud_names <- names(cloud)
  
  missing_in_cloud <- local_names[!(local_names %in% cloud_names)]
  missing_in_local <- cloud_names[!(cloud_names %in% local_names)]
  
  size_mismatches = list()
  common_names <- local_names[local_names %in% cloud_names]
  
  for (p in common_names) {
    if (local[[p]] != cloud[[p]]) {
      size_mismatches[[p]] <- list(local = local[[p]], cloud = cloud[[p]])
    }
  }
  
  clean_ext <- gsub("^\\.*", "", video_extension)
  
  cat("\n=====================================\n")
  cat("        COMPARISON RESULTS           \n")
  cat("=====================================\n")
  cat(sprintf("Local drive count: %d %s files\n", length(local), clean_ext))
  cat(sprintf("GCP Bucket count:  %d %s objects\n", length(cloud), clean_ext))
  cat("-------------------------------------\n")
  
  if (length(missing_in_cloud) == 0 && length(missing_in_local) == 0 && length(size_mismatches) == 0) {
    cat("🎉 SUCCESS! Every file matches perfectly in name and size.\n")
  } else {
    if (length(missing_in_cloud) > 0) {
      cat(sprintf("❌ Missing in Cloud (%d):\n", length(missing_in_cloud)))
      limit <- min(10, length(missing_in_cloud))
      for (i in 1:limit) cat(sprintf("  - %s\n", missing_in_cloud[i]))
      if (length(missing_in_cloud) > 10) cat(sprintf("  ... and %d more.\n", length(missing_in_cloud) - 10))
    }
    
    if (length(missing_in_local) > 0) {
      cat(sprintf("⚠️ Extra in Cloud / Missing Locally (%d):\n", length(missing_in_local)))
      limit <- min(10, length(missing_in_local))
      for (i in 1:limit) cat(sprintf("  - %s\n", missing_in_local[i]))
      if (length(missing_in_local) > 10) cat(sprintf("  ... and %d more.\n", length(missing_in_local) - 10))
    }
    
    if (length(size_mismatches) > 0) {
      cat(sprintf("❌ Byte Size Mismatches (%d):\n", length(size_mismatches)))
      mismatch_names <- names(size_mismatches)
      limit <- min(10, length(mismatch_names))
      for (i in 1:limit) {
        p <- mismatch_names[i]
        cat(sprintf("  - %s (Local: %.0f bytes, Cloud: %.0f bytes)\n", 
                    p, size_mismatches[[p]]$local, size_mismatches[[p]]$cloud))
      }
      if (length(size_mismatches) > 10) cat(sprintf("  ... and %d more.\n", length(size_mismatches) - 10))
    }
  }
}

# =============================================================================
# MAIN RUNNER
# =============================================================================

main <- function() {
  config_path <- parse_args()
  config <- load_config(config_path)
  
  cloud_inventory <- get_cloud_manifest(
    bucket_path = config$gcp_bucket_path,
    extension = config$video_extension
  )
  
  prefix <- extract_gcp_prefix(config$gcp_bucket_path)
  
  local_inventory <- get_local_manifest(
    local_path = config$output_directory,
    prefix = prefix,
    extension = config$video_extension
  )
  
  compare_inventories(
    local = local_inventory, 
    cloud = cloud_inventory, 
    video_extension = config$video_extension
  )
}

main()